src/inc/til/env.h

```diff
@@ -0,0 +1,515 @@
+                                        data = data.substr(0, data.size() - 1);
+                                    }
+                                }
+
+                                if (!data.empty())
+                                {
+                                    if (is_path_var(valueName))
+                                    {
+                                        concat_var(valueName, std::wstring{ data });
+                                    }
+                                    else
+                                    {
+                                        set_user_environment_var(valueName, std::wstring{ data });
+                                    }
+                                }
+                            }
+
+                            valueName.resize(maxValueNameSize);
+                            valueData.resize(maxValueDataSize);
+                            valueNameSize = maxValueNameSize;
+                            valueDataSize = maxValueDataSize;
+                            index++;
+                        }
+                        index = 0;
+                    }
+                }
+            }
+        }
+
+        std::wstring expand_environment_strings(std::wstring input)
+        {
+            // TODO: this should be replacing from ourselves, not from the OS
+            return wil::ExpandEnvironmentStringsW<std::wstring, 256>(input.data());
+        }
+
+        void set_user_environment_var(std::wstring var, std::wstring value)
+        {
+            value = expand_environment_strings(value);
+            value = check_for_temp(value);
+            save_to_map(var, value);
+        }
+
+        void concat_var(std::wstring var, std::wstring value)
+        {
+            if (const auto existing = _envMap.find(var); existing != _envMap.end())
+            {
+                // If it doesn't already trail with a ;, add one.
+                // Otherwise, just take advantage of the one it has.
+                if (existing->second.back() != L';')
+                {
+                    existing->second.append(L";");
+                }
+                existing->second.append(value);
+                save_to_map(var, existing->second);
+            }
+            else
+            {
+                save_to_map(var, value);
+            }
+        }
+
+        void save_to_map(std::wstring var, std::wstring value)
+        {
+            if (!var.empty() && !value.empty())
+            {
+                _envMap.insert_or_assign(var, value);
+            }
+        }
+
+        static constexpr std::wstring_view temp{ L"temp" };
+        static constexpr std::wstring_view tmp{ L"tmp" };
+        std::wstring check_for_temp(std::wstring_view input)
+        {
+            if (!_wcsicmp(input.data(), temp.data()) ||
+                !_wcsicmp(input.data(), tmp.data()))
+            {
+                return til::details::wil_env::GetShortPathNameW<std::wstring, 256>(input.data());

```

Comment:
it looks like this will expand any variable whose _value_ is identical to `temp` to.. `temp`

Reply:
Yeah - I had twigged this, I just haven't had time to fix it - I'm thinking this method probably needs to be given the environment variable name too!